executable/build.go

```diff
@@ -37,9 +43,84 @@ type Build struct {
 func (b Build) Build(context libcnb.BuildContext) (libcnb.BuildResult, error) {
 	result := libcnb.NewBuildResult()
 
-	m, err := libjvm.NewManifest(context.Application.Path)
-	if err != nil {
-		return libcnb.BuildResult{}, fmt.Errorf("unable to read manifest in %s\n%w", context.Application.Path, err)
+	// Check if there is a top-level META-INF/MANIFEST.MF
+	manifest := filepath.Join(context.Application.Path, "META-INF", "MANIFEST.MF")
+	manifestExists := true
+	if _, err := os.Stat(manifest); errors.Is(err, os.ErrNotExist) {
+		manifestExists = false
+	}
+
+	var allJarFiles []string
+	m := properties.NewProperties()
+
+	if manifestExists {
+		var err error
+		m, err = libjvm.NewManifest(context.Application.Path)
+		if err != nil {
+			return libcnb.BuildResult{}, fmt.Errorf("unable to read manifest in %s\n%w", context.Application.Path, err)
+		}
+	} else {
+		// loop through the JAR files, collect them for the classpath and pick the first Main-Class
+		items, err := ioutil.ReadDir(context.Application.Path)
+		if err != nil {
+			return libcnb.BuildResult{}, fmt.Errorf("unable to read files in %s\n%w", context.Application.Path, err)
+		}
+
+		mainClassFound := false
+		for _, item := range items {
+			// make sure it is a file
+			if item.IsDir() {
+				continue
+			}
+
+			// make sure it is a JAR file
+			fileName := item.Name()
+			if !strings.HasSuffix(strings.ToLower(fileName), ".jar") {
+				continue
+			}
+
+			// append it to the list of all JAR files
+			jarFileName := filepath.Join(context.Application.Path, fileName)
+			allJarFiles = append(allJarFiles, jarFileName)
+
+			// if we already found a manifest with Main-Class, we can continue with the next file
+			if mainClassFound {
+				continue
+			}
+
+			// open the JAR file
+			z, err := zip.OpenReader(jarFileName)
+			if err != nil {
+				return libcnb.BuildResult{}, fmt.Errorf("unable to read file %s\n%w", jarFileName, err)
+			}
+			defer z.Close()
+
+			// look for the MANIFEST
+			for _, f := range z.File {
```

Comment:
Is the loop necessary? Wouldn't `z.Open()` work?

Reply:
Good catch.